fix: handle 404 no found and 409 conflict errrors during subscriber operations

gatici opened the pull request:
# Description

The backend started to throw `Not Found Error` if subscriber does not exist and `Conflict Error` if subscriber already Exist within the following PRs:
https://github.com/omec-project/webconsole/pull/318
https://github.com/omec-project/webconsole/pull/317

This PR handles those errors in the frontend which are throwed by backend.

# Checklist:

- [ ] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [ ] I have performed a self-review of my own code.
- [ ] I have made corresponding changes to the documentation.
- [ ] I have added tests that validate the behaviour of the software.
- [ ] I validated that new and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.

patriciareinoso commented on utils/subscriberOperations.tsx:
```diff
@@ -221,7 +262,6 @@ export async function editSubscriber({
     if (!exists){
       throw new OperationError("Subscriber does not exist.");
     }
-    await apiPostSubscriber(subscriberData.rawImsi, subscriberData, token);
```
since we are no longer using the the whole information in `subscriberData` we could change the input parameters in EditSubscriberArgs to replace subscriberData with only the imsi.

gatici replied:
Done.

patriciareinoso commented on utils/subscriberOperations.tsx:
```diff
@@ -185,7 +221,7 @@ export async function createSubscriber({
   try {
```
here in createSubscriber we are not leveraging the new 409 error.
we can remove the call to doesSubscriberExist and trust the apiPostSubscriber call.

gatici replied:
Done

patriciareinoso commented on utils/subscriberOperations.tsx:
```diff
@@ -196,6 +232,11 @@ export async function createSubscriber({
     await apiPostDeviceGroup(deviceGroupName, existingDeviceGroupData, token);
 
   } catch (error) {
+    if (is409ConflictError(error)) {
```
I would catch the 409 error just around the call to apiPostSubscriber, like this we are sure that this was the one throwing this exception and know that the problem is with the subscriber creation. 
I would keep the same behavior as before: raising an Operation error with the appropriate message so that is shown in the modal.

gatici replied:
Done

patriciareinoso commented on utils/subscriberOperations.tsx:
```diff
@@ -144,6 +171,10 @@ async function getSubscriberAuthData(ueId: string, token: string): Promise<Subsc
     }
     return subscriberAuthData;
   } catch (error) {
+    if (is404NotFoundError(error)) {
+      console.error(`Subscriber with IMSI ${ueId.split("-")[1]} does not exist.`);
+      throw error
+    }
```
```suggestion
```
all of this seems redundant

patriciareinoso reviewed (changes requested):
The idea of using exception types in the NMS was to be able show appropriate messages to the user.  We would not want to show something including coding errors to the user.

At the moment the modals catch the following errors:
1. 401 error  -> to log out the user 
2. OperationError  -> These are the errors that are meant to be shown to the user. They are displayed in the modal and are errors related to something the user can fix. For example: subscriber imsi already exist.
3. Any other error -> diplay "An unexpected error occurred."

```
if (is401UnauthorizedError(error)) {
            auth.logout();
        } else if (error instanceof OperationError) {
          setApiError(error.message);
        } else {
          setApiError("An unexpected error occurred.");
        }
```

The tables catch:
1. 401 error -> log out
2. In any other case it displays a generic message like "Failed to retrieve users"

```
if (queries.some(q => q.status === "error")) {
    if (queries.some(q => is401UnauthorizedError(q.error))) {
      auth.logout();
    }
    return (<><ErrorNotification error={"Failed to retrieve network slices."} /></>);
  }
 ```

gatici commented:
> The idea of using exception types in the NMS was to be able show appropriate messages to the user. We would not want to show something including coding errors to the user.
> 
> At the moment the modals catch the following errors:
> 
>     1. 401 error  -> to log out the user
> 
>     2. OperationError  -> These are the errors that are meant to be shown to the user. They are displayed in the modal and are errors related to something the user can fix. For example: subscriber imsi already exist.
> 
>     3. Any other error -> diplay "An unexpected error occurred."
> 
> 
> ```
> if (is401UnauthorizedError(error)) {
>             auth.logout();
>         } else if (error instanceof OperationError) {
>           setApiError(error.message);
>         } else {
>           setApiError("An unexpected error occurred.");
>         }
> ```
> 
> The tables catch:
> 
>     1. 401 error -> log out
> 
>     2. In any other case it displays a generic message like "Failed to retrieve users"
> 
> 
> ```
> if (queries.some(q => q.status === "error")) {
>     if (queries.some(q => is401UnauthorizedError(q.error))) {
>       auth.logout();
>     }
>     return (<><ErrorNotification error={"Failed to retrieve network slices."} /></>);
>   }
> ```

@patriciareinoso Thank you for the review and detailed explanation. 
I addressed all the comments.

gatici commented on utils/errors.tsx:
```diff
@@ -24,6 +24,11 @@ export function is404NotFoundError(error: Error | unknown): boolean {
 }
 
+export function is409ConflictError(error: Error | unknown): boolean {
+  return (error instanceof WebconsoleApiError && error.status === 409);
```
Done.

gatici commented on utils/subscriberOperations.tsx:
```diff
@@ -59,9 +75,19 @@ async function apiPostSubscriber(rawImsi: string, subscriberData: SubscriberAuth
+    if (response.status === 409) {
+      throw new WebconsoleApiError(409, "Subscriber already exists");
+    }
+
```
Done

gatici commented on utils/subscriberOperations.tsx:
```diff
@@ -33,13 +32,30 @@ async function apiGetSubscriber(rawImsi: string, token: string): Promise<Respons
     throw new OperationError(`Error getting subscriber: Invalid imsi provided ${rawImsi}.`);
```
Done

gatici commented:
> we need a new webconsole tag and update it in the rock

A new tag `1.8.4`  is created and used in this PR.

patriciareinoso reviewed (commented):
Looks so much cleaner :)

gatici commented on utils/subscriberOperations.tsx:
```diff
@@ -130,12 +129,16 @@ async function getSubscriberAuthData(ueId: string, token: string): Promise<Subsc
-    const subscriber = await response.json();
     if (!response.ok) {
-      throw new WebconsoleApiError(response.status, subscriber.error);
+      console.error(`Error retrieving subscriber data for ${rawImsi}: ${response.status}`);
```
Removed.

gatici commented on utils/subscriberOperations.tsx:
```diff
@@ -163,9 +166,14 @@ export async function filterSubscribers(subscribers: string[], token: string): P
+    if (is404NotFoundError(error)) {
+      return false;
+    }
+
```
OK. Fixed.
